### DISCUSSION TITLE
LDM notes for August 9, 2023

### DISCUSSION THREAD

**Discussion by MadsTorgersen** [dotnet/csharplang#7457] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-08-09.md

## Agenda

- Lambdas with explicit return types
- Target typing of collection expressions to core interfaces
- Loosening requirements for collection builder methods

**Comment by TonyValenti** [comment]
If it is not already enabled, it would be great if interfaces could also be tagged with the collection builder attribute.

**Reply by CyrusNajmabadi** [reply]
They can be @TonyValenti . This is how types like `IImutableList<T>` work with this feature.

**Comment by sab39** [comment]
The current C#12 proposal, as I understand it, is that collection expressions have no natural type. Does that mean that they won't work with `foreach` out of the box? That would be pretty disappointing, if so, considering that while we may not know the collection type, we _do_ know the iteration type, and so we know exactly what `foreach` should do, for _any_ well-behaved collection type.

Further, since the actual collection can't be observed, the compiler doesn't need to pick or even use a collection type at all - it could optimize
```cs
foreach (var num in [1, 2, 3])
{
  func(num);
}
```
by unrolling it directly into
```cs
func(1);
func(2);
func(3);
```

Am I right that without a natural type we don't get `foreach` for free? If so, I propose that a special case rule be added to permit collection expressions in `foreach` regardless, because IMO it would be very surprising and frustrating to users if that didn't work.

**Reply by HaloFour** [reply]
I don't think the one precludes the other, but I do believe that the goal is that by C# 13 that collection literals would support a natural type, and the team has expressed optimizing away the collection in those cases where it is not observable, such as with `foreach` loops.

**Reply by sab39** [reply]
Yeah, the plan for C#13 covers all the things I was concerned about - my contention is that it would be very surprising if `foreach (var num in [1, 2, 3])` didn't work as expected _in C#12_ due to the lack of a natural type. The rule could be something as simple as treating a collection expression in a `foreach` as if it were target-typed to `IEnumerable<>`, for example.

**Reply by CyrusNajmabadi** [reply]
> Does that mean that they won't work with foreach out of the box?

No.  We are discussing ways to make them work with `foreach` without a natural type.  for example by target-typing them to `ReadOnlySpan<ForeachElementType>`.  If so, then  `foreach (bool b in [true, false])` would work (and be efficient).

For the `foreach (var v in [true, false])` that would continue not working, as we would have no inferrable type here.

**Reply by CyrusNajmabadi** [reply]
> by unrolling it directly into

It is not at all clear that unrolling is an optimization.  We leave those determinations to the runtime (which can do things like apply PGO techniques to see if it is wise to do so).

**Reply by sab39** [reply]
> For the `foreach (var v in [true, false])` that would continue not working, as we would have no inferrable type here.

Surely if you're stipulating `ReadOnlySpan<>` then you could infer the element type the same way as you would for something like this, which I'm pretty sure already works with the current C#12 proposal?

```cs
void doThing<T>(ReadOnlySpan<T> arg);
doThing([true, false]);
```

> It is not at all clear that unrolling is an optimization

That's very fair - my intent was only to suggest it be _allowed_, not that it would necessarily be beneficial, but you're right that it makes more sense to leave that kind of thing to the JIT.

**Reply by CyrusNajmabadi** [reply]
> Surely if you're stipulating ReadOnlySpan<> then you could infer the element type the same way as you would for something like this, which I'm pretty sure already works with the current C#12 proposal?

Yes, that's a reasonable thing to consider.  Thanks!

**Reply by sab39** [reply]
> We are discussing ways to make them work with `foreach` without a natural type. for example by target-typing them to `ReadOnlySpan<ForeachElementType>`.

Also, `ReadOnlySpan<>` wouldn't be optimal in all cases - particularly unknown-length literals. Consider
```cs
foreach (var i in [..Enumerable.Range(1, 1000), ..Enumerable.Range(2000, 3000)])
```

Seems like it would be better to leave the compiler flexibility about what type to pick rather than specifying something concrete. In much the same way as target-typing to `IEnumerable<T>` guarantees you'll get a type that implements that interface but allows the compiler complete freedom in what specific type to use, `foreach` could guarantee you'll get something that (behaves as if it) implements the pattern of methods `foreach` expects, but leave it up to the compiler what the actual type is.

**Reply by CyrusNajmabadi** [reply]
> Also, ReadOnlySpan<> wouldn't be optimal in all cases - particularly unknown-length literals. Consider

Target-typign has no relation to what is emitted.  It just means how we determine applicability/conversions.  The actual emit logic can be optimized however we feel is appropriate based on input types.

**Reply by AartBluestoke** [reply]
"Target-typing has no relation to what is emitted."

is it possible that knowing that the current implicit list is being compiled for a `foreach (var i in [true,false])` 
could produce different output to a call `foo([true,false])` despite the same implicit list? - having the context in which the implicit list appear (ie the type it is targeted against) influence what the compiler does?

and would such things be a matter for the language, or just an implementation detail?

<edited for clarity>

**Reply by CyrusNajmabadi** [reply]
@AartBluestoke sorry, i'm not quite sure what your question is.  Can you clarify?  Thanks!

**Reply by CyrusNajmabadi** [reply]
> having the context in which the implicit list appear (ie the type it is targeted against) influence what the compiler does?

Absolutely. The target very much will affect what the compiler does. 

> and would such things be a matter for the language, or just an implementation detail?

The language determines what is legal and what semantics something has. The compiler is free to implement things how it wants as long as what it does isn't *observably* different from the required semantics.
